common/service/host/kcs.c

```diff
@@ -149,9 +150,44 @@ void kcs_read(void *arvg0, void *arvg1, void *arvg2)
 				memcpy(&bridge_msg.data[0], &ibuf[2], rc);
 			}
 
-			status = ipmb_send_request(&bridge_msg, IPMB_inf_index_map[BMC_IPMB]);
-			if (status != IPMB_ERROR_SUCCESS) {
-				LOG_ERR("kcs_read_task send to BMC fail status: 0x%x", status);
+
+			// Check BMC communication interface if use IPMB or not
+			if (!pal_is_interface_use_ipmb(IPMB_inf_index_map[BMC_IPMB])) {
+				int ret = 0;
+
+				// Send request to MCTP/PLDM thread to ask BMC
+				ret = pldm_send_ipmi_request(&bridge_msg);
+				if (ret < 0) {
+					LOG_ERR("kcs_read_task send to BMC fail");
+				}
+
+				uint8_t *kcs_buff;
+				kcs_buff = malloc(KCS_BUFF_SIZE * sizeof(uint8_t));
+				if (kcs_buff == NULL) {
+					LOG_ERR("Memory allocation failed");
+					continue;
+				}
+
+				kcs_buff[0] = (bridge_msg.netfn | BIT(0)) << 2;
+				kcs_buff[1] = bridge_msg.cmd;
+				kcs_buff[2] = bridge_msg.completion_code;
+
+				// Write MCTP/PLDM response to KCS
+				if (bridge_msg.data_len > 0) {
```

Comment:
Ideally we would add an iovec-style kcs_write routine in order to reduce the amount of memcpy and buffers we need to allocate.

Reply:
Could you explain iovec-style routine in more detail? 
We're not sure we understand it.
And could we update this next commit?
Thank you.

Reply:
Here is a reference on vectored I/O support: https://dius-au.medium.com/good-design-from-the-ground-up-how-i-o-library-design-can-make-or-break-iot-solutions-49a66bc6c5b2 which Patrick is referencing on how to relieve memory pressure/usage.